base/loading.jl

```diff
@@ -2,6 +2,52 @@
 
 # Base.require is the implementation for the `import` statement
 
+# Generic case-sensitive version of isfile
+function isfile_casesensitive_slow(path)
+    isfile(path) || return false
+    dir, filename = splitdir(path)
+    any(readdir(dir) .== filename)
+end
+
+isfile_casesensitive(path) = isfile_casesensitive_slow(path)
+
+@windows_only function isfile_casesensitive(path)
+    isfile(path) || return false
+    longpath(path) == path
+end
+
+@linux_only isfile_casesensitive(path) = isfile(path)
+
```

Comment:
@malmaud's code below already avoids walking the directory on OS X. `fcntl` with `F_GETPATH` seems like another possibility, although I'm not sure what it does if the file is a symlink, or if there are hard links.